### ISSUE TITLE
Feature Request: Keyboard navigation inside the popup

### ISSUE THREAD

**Issue by girst** [uBlockOrigin/uMatrix-issues#197]
### Prerequisites

- [x] I performed a cursory search of the issue tracker to avoid opening a duplicate issue
    - Your issue may already be reported.
    - [x] I also searched the existing issues at <https://github.com/gorhill/uMatrix/issues>
- [x] This is not a support issue or a question
    - Support issues and questions are handled at [/r/uMatrix](https://old.reddit.com/r/uMatrix/)
- I tried to reproduce the issue when...
    - [ ] uMatrix is the only extension
    - [ ] uMatrix with default lists/settings
    - [ ] using a new, unmodified browser profile
    - **note**: *above not applicable; feature request*
- [x] I am running the latest version of uMatrix
- [x] I checked the [documentation](https://github.com/gorhill/uMatrix/wiki) to understand that the issue I report is not a normal behavior
- [ ] I used the logger to rule out that the issue is caused by my ruleset **note**: *not applicable; feature request*

### Description

As a keyboard centric user, I would like to be able to select/whitelist/blacklist cells/rows/columns using the keyboard (preferably hjkl (vi-like) and cursor keys). I've hacked together a quick proof of concept; its source (.diff.txt) and a video (.zip) of it in action is attached.
[umatrix-keyboard.diff.txt](https://github.com/uBlockOrigin/uMatrix-issues/files/3799613/umatrix-keyboard.diff.txt)
[demo.zip](https://github.com/uBlockOrigin/uMatrix-issues/files/3799628/demo.zip)

I'm happy to put together a proper patch for this, if a) there's interest and b) if I can get some pointers about how to best move between cells (the current approach is a bit cumbersome).

### A specific URL where the issue occurs

**note**: *not applicable; feature request*

### Steps to Reproduce

1. open the uMatrix popup
2. highlight cells using the keyboard
3. hit spacebar/return to toggle its state

### Ruleset

**note**: *not applicable; feature request*

### Supporting evidence

**note**: *not applicable; feature request*. however, see the attached patch and video

### Your environment

* uMatrix version: 1.4.0
* Browser Name and version: firefox-70.0-1.fc30.x86_64
* Operating System and version: fedora 30

**Comment by gwarser** [comment]
Blocked by https://github.com/uBlockOrigin/uMatrix-issues/issues/37 ?

**Comment by uBlock-user** [comment]
sounds like a duplicate of that one

**Comment by gorhill** [comment]
It's different. He wants keyboard-based navigation in the popup panel, while `browser.commands` is for issuing commands unrelated to the popup panel.

**Comment by girst** [comment]
On Fri, Nov 01, 2019 at 03:01:31PM -0700, Raymond Hill wrote:
>It's different. He wants keyboard-based navigation in the popup panel

correct! #37 would be quite useful to have working too, but VimFx 
provides me with a key combo already (`eb` to select browser elements).

i'll re-state my intention to implement this myself, as long as there's 
a reasonable chance for it to get merged. I will need some pointers on 
how to best implement navigation between cells, though.

**Comment by girst** [comment]
@gorhill: I've implemented this now. I'm quite happy with how it turned out, but as I'm not really a JavaScript developer, a code review is probably not a bad idea ;-). The commit message contains my implementation notes.

I'm not allowed to open a pull request, but you can check out the feature branch at https://github.com/girst/uMatrix/tree/pr-keyboardnav

**Comment by gorhill** [comment]
I am not having much time for this right now, so for now just keep using your version -- I do not commit often to uMatrix anymore, so you shouldn't have to worry about your changes. Eventually I may get more time to go through this.

**Comment by girst** [comment]
On Sun, Nov 03, 2019 at 07:44:11AM -0800, Raymond Hill wrote:
>I am not having much time for this right now, so for now just keep 
>using your version -- I do not commit often to uMatrix anymore, so you 
>shouldn't have to worry about your changes. Eventually I may get more 
>time to go through this.

Thanks for the response. There's no rush! I'll keep forward-porting that 
as long as necessary, but I do hope this can eventually get merged (I'd 
prefer using the (properly signed) upstream version and of course not 
having to apply local patches).

**Comment by girst** [comment]
[uMatrix-keyboardnav-xpi.zip](https://github.com/uBlockOrigin/uMatrix-issues/files/3828550/uMatrix-keyboardnav-xpi.zip)

I've now implemented a checkbox in the settings where this mode is also documented. a (unsigned, renamed) firefox addon is attached, for those interested.

**Comment by girst** [comment]
@gorhill: seeing htat you pushed some commits the last few days, can you please consider pulling from https://github.com/girst/uMatrix/tree/pr-keyboardnav (i'm not allowed to open pull requests on the repo, as I'm not a "previous contributor")?

thanks

**Comment by elig0n** [comment]
Will @girst patches be admitted? Does @girst maintain his patch up-to-date?

**Comment by girst** [comment]
On Sun, Apr 26, 2020 at 06:02:24AM -0700, Eliran Gonen wrote:
>Will @girst patches be admitted?

I hope so, but I haven't heard anything from gorhill :(

> Does @girst maintain his patch up-to-date?

I do. Just rebased on top of gorhill/master (applied automatically); 
thanks for the reminder.  Please ping me if it requires rebasing in the 
future, too.
